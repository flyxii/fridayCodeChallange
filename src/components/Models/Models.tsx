/**
 * Component to fetch the list of models and display the return data to end user
 * 
 */

import React, { useContext } from "react";
import useSWR from "swr";
import { AppContext } from "store/context";
import { Action } from "store/type";
import Loading from "libs/images/Loading.svg";
import styles from "./Models.module.scss";

const Models: React.FC = () => {
  /**
   * Connect to the store
   */
  const { state, dispatch } = useContext(AppContext);
  const { make, model } = state;

  /**
   * Fetch models data from api
   */
  const { data, error } = useSWR<string[], Error>(`http://localhost:8080/api/models?make=${make}`);

  /**
   * Show empty box if make is falsy
   */
  if (!make) return <div className={styles.error} data-testid="empty-wrapper"></div>;

  /**
   * Show loading icon if waiting for data to come or error return from api
   */
  if (error || !data)
    return (
      <div className={styles.error}>
        <img src={Loading} className={styles.loading} alt="Loading" data-testid="loading" />
      </div>
    );

  /**
   * Show error message if no data return from api
   */
  if (data.length === 0) return (
    <div className={styles.error} data-testid="no-data">
      No data found
    </div>
  );

  /**
   * Show the list of models
   */
  return (
    <div className={styles.wrapper}>
      <div className={styles.title}>Models</div>
      <div className={styles.scrollWrapper}>
        {data.map((item, i) => (
          <div
            className={model === item ? styles.model + " " + styles.selected : styles.model}
            key={i}
            onClick={() => {
              dispatch({ type: Action.UPDATE_MODEL, payload: item });
            }}>
            {item}
          </div>
        ))}
      </div>
    </div>
  );
};

export default Models;
