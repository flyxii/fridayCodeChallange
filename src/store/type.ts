export interface State {
  make: string;
  model: string;
}

export type UPDATE_MAKE = { type: Action.UPDATE_MAKE; payload: string };
export type UPDATE_MODEL = { type: Action.UPDATE_MODEL; payload: string };

export enum Action {
  UPDATE_MAKE,
  UPDATE_MODEL,
}

export type Actions = UPDATE_MAKE | UPDATE_MODEL;
