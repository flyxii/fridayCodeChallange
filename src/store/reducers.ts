import { State, Actions, Action } from "./type";

export const reducer = (state: State, action: Actions): State => {
  switch (action.type) {
    case Action.UPDATE_MAKE:
      return { ...state, make: action.payload };
    case Action.UPDATE_MODEL:
      return { ...state, model: action.payload };
    default:
      return state;
  }
};
