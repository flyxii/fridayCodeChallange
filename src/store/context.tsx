/**
 * The useContext hook to create a store that could share data among components
 */

import React, { createContext, useReducer } from "react";
import { reducer } from "./reducers";
import { State, Actions } from "./type";

const initialState: State = {
  make: "",
  model: "",
};

const AppContext = createContext<{
  state: State;
  dispatch: React.Dispatch<Actions>;
}>({
  state: initialState,
  dispatch: () => null,
});

const AppProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return <AppContext.Provider value={{ state, dispatch }}>{children}</AppContext.Provider>;
};

export { AppContext, AppProvider };
