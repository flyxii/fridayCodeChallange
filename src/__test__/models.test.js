import * as React from "react";
import { SWRConfig } from "swr";
import fetch from "jest-fetch-mock";
import { render, waitFor } from "@testing-library/react";
import { AppContext } from "store/context";
import Models from "components/Models";
import fetcher from "libs/fetcher";

const AppContextWithoutValue = () => (
  <AppContext.Provider value={{ state: { make: "", model: "" } }}>
    <SWRConfig value={{ dedupingInterval: 0, shouldRetryOnError: false, fetcher: fetcher }}>
      <Models />
    </SWRConfig>
  </AppContext.Provider>
);

const AppContextWithValue = () => (
  <AppContext.Provider value={{ state: { make: "Ford", model: "" } }}>
    <SWRConfig value={{ dedupingInterval: 0, shouldRetryOnError: false, fetcher: fetcher }}>
      <Models />
    </SWRConfig>
  </AppContext.Provider>
);

describe("<Models />", () => {
  beforeEach(() => {
    fetch.resetMocks();
    fetch.doMock();
    jest.useFakeTimers();
  });
  afterEach(() => {
    jest.useRealTimers();
  });

  it("renders without crashing", async () => {
    const { container, asFragment } = render(<AppContextWithValue />);
    await waitFor(() => {
      expect(container.firstChild).toBeInTheDocument();
    });

     expect(asFragment()).toMatchSnapshot();
  });

  it("shows empty wrapper if make value is falsy", async () => {
    fetch.mockReject(() => Promise.reject("server is down"));
    const { getByTestId, asFragment } = render(<AppContextWithoutValue />);

    await waitFor(() => {
      const emptyWrapper = getByTestId(/empty-wrapper/i);
      expect(emptyWrapper).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("shows laoding icon if fetch reject", async () => {
    fetch.mockReject(() => Promise.reject("server is down"));
    const { getByTestId, asFragment } = render(<AppContextWithValue />);

    await waitFor(() => {
      const loading = getByTestId(/loading/i);
      expect(loading).toBeInTheDocument();
      expect(fetch).toHaveBeenCalledTimes(1);
      expect(fetch).toHaveBeenCalledWith("http://localhost:8080/api/models?make=Ford", undefined);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("shows loading icon in loading stage", async () => {
    fetch.mockResponseOnce("");
    const { getByTestId, asFragment } = render(<AppContextWithValue />);

    await waitFor(() => {
      const loading = getByTestId(/loading/i);
      expect(loading).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("shows message if fetch success with empty array", async () => {
    const mockResponse = [];

    fetch.mockResponseOnce(JSON.stringify(mockResponse));
    const { getByTestId, asFragment } = render(<AppContextWithValue />);

    await waitFor(() => {
      const loading = getByTestId(/no-data/i);
      expect(loading).toBeInTheDocument();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows data if fetch success", async () => {
    const mockResponse = ["Explorer", "Fiesta"];

    fetch.mockResponseOnce(JSON.stringify(mockResponse));
    const { findAllByText, asFragment } = render(<AppContextWithValue />);

    const explorer = await findAllByText("Explorer");
    const fiesta = await findAllByText("Fiesta");

    expect(explorer).toHaveLength(1);
    expect(fiesta).toHaveLength(1);
    expect(asFragment()).toMatchSnapshot();
  });
});
