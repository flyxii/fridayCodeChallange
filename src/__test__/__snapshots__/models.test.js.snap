// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Models /> renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="error"
  >
    <img
      alt="Loading"
      class="loading"
      data-testid="loading"
      src="Loading.svg"
    />
  </div>
</DocumentFragment>
`;

exports[`<Models /> shows data if fetch success 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <div
      class="title"
    >
      Models
    </div>
    <div
      class="scrollWrapper"
    >
      <div
        class="model"
      >
        Explorer
      </div>
      <div
        class="model"
      >
        Fiesta
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Models /> shows empty wrapper if make value is falsy 1`] = `
<DocumentFragment>
  <div
    class="error"
    data-testid="empty-wrapper"
  />
</DocumentFragment>
`;

exports[`<Models /> shows laoding icon if fetch reject 1`] = `
<DocumentFragment>
  <div
    class="error"
  >
    <img
      alt="Loading"
      class="loading"
      data-testid="loading"
      src="Loading.svg"
    />
  </div>
</DocumentFragment>
`;

exports[`<Models /> shows loading icon in loading stage 1`] = `
<DocumentFragment>
  <div
    class="error"
  >
    <img
      alt="Loading"
      class="loading"
      data-testid="loading"
      src="Loading.svg"
    />
  </div>
</DocumentFragment>
`;

exports[`<Models /> shows message if fetch success with empty array 1`] = `
<DocumentFragment>
  <div
    class="error"
    data-testid="no-data"
  >
    No data found
  </div>
</DocumentFragment>
`;
